### ISSUE TITLE
Compress is actually CompressStream

### ISSUE THREAD

**Issue by MattPryze** [gibbed/XCompression#2]
Digging into this, I realized that XMemCompress is functionally XMemCompressStream with a different name. This is identified in the result of the data as well as the fact that the 5th argument is a reference to an int and not just a normal int. Compressed data ends up missing the last chunk since it's being held in a buffer somewhere and isn't being flushed.

I couldn't locate a flush function or anything similar so I'm not sure where to go from here.

**Comment by gibbed** [comment]
That's interesting, I don't remember having trouble when I was actively using my XCompression to do compression. Do you have sample code that produces an incomplete compressed output?

**Comment by MattPryze** [comment]
I do. I translated your library to C++, however the issue was present in both your library and the C++ library. So the sample code I have would be C++.

For reference, here is a snippet taken from xcompress.h that shows there are 2 different compression implementations. Compress and CompressStream. The compress found in your library is functionally identical to CompressStream.


```
XBOXAPI
HRESULT 
WINAPI 
XMemCompress(
    __in    XMEMCOMPRESSION_CONTEXT         Context,
    __out_bcount_opt( *pDestSize ) VOID*    pDestination,
    __inout SIZE_T*                         pDestSize,
    __in_bcount( SrcSize ) CONST VOID*      pSource,
    __in    SIZE_T                          SrcSize
    );

XBOXAPI
HRESULT 
WINAPI 
XMemCompressStream(
    __in    XMEMCOMPRESSION_CONTEXT         Context,
    __out_bcount_opt( *pDestSize ) VOID*    pDestination,
    __inout SIZE_T*                         pDestSize,
    __in_bcount( *pSrcSize ) CONST VOID*    pSource,
    __inout SIZE_T*                         pSrcSize
    );
```

The flag of 1 sent in to CreateCompressionContext also stands for `XMEMCOMPRESS_STREAM` as indicated in xcompress.h and when it's set to 0, it leads to a crash.

Here is an example of my implementation using the C++ equivalent.
```
int XCompress(char *CompressedBuffer, char *UncompressedBuffer, int UncompressedSize)
{
    COMPRESSIONSETTINGS settings = { 1, DEFAULT_WINDOW_SIZE, DEFAULT_CHUNK_SIZE };

    void *context;
    int r;
    r = XMemCreateCompressionContext(1, &settings, 1, &context);
    SIZE_T newLen = UncompressedSize * 2;
    r = XMemCompress(context, CompressedBuffer, &newLen, UncompressedBuffer, &UncompressedSize);
    XMemDestroyCompressionContext(context);
    return newLen;
}
```

Starting with a compressed file of size 3.96Mb, decompressing it yields a full result. All data is there. When compressing it back down, it's only 3.91Mb. Decompressing THAT result shows the trailing chunk of data is not there. Also, when compressing small amounts of data, the Compress function doesn't return anything.

It's definitely a streaming function, I just don't know how to use it without a flush.

I also tried every version of XnaNative. Same result.

**Comment by MattPryze** [comment]
For clarification, using xcompress's XMemCompressStream yields the same exact file size changes and same data loss.

**Comment by gibbed** [comment]
Looking up my old code where I use XCompression for compression, you pass `0` in `inputSize` to flush.

**Comment by gibbed** [comment]
Which actually reveals I have a long outstanding commit I never pushed to change the bounds checking on the `Compress` wrapper that didn't make the jump from SVN to GitHub.

**Comment by MattPryze** [comment]
Hmmm when I set `UncompressedSize` to 0, the output is empty. Do you have some example code?

**Comment by gibbed** [comment]
@MRG95 Yeah, give me a bit to make some fix pushes on this repo to address the outdated code and modernization in general.

**Comment by MattPryze** [comment]
I just got it working. You are a lifesaver! Thanks so much.

**Comment by gibbed** [comment]
@MRG95 Great to hear, I'll still fix up XCompression regarding flushing the compression output.
